### PR TITLE
[AIRFLOW-2824] - Add config to disable default conn creation

### PR DESCRIPTION
**Pull request by andscoop**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-2824
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR gives users a way to avoid recreating default connections when running initDB via the airflow.cfg file

Before merging - I wanted feedback from the maintainers on whether or not they feel this is the way to approach resolving this issue.

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
Documented in config code comments

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: airflow/utils/db.py [hunk 1]
```diff
@@ -286,6 +284,16 @@ def initdb(rbac=False):
             conn_id='cassandra_default', conn_type='cassandra',
             host='cassandra', port=9042))
 
+
+def initdb(rbac=False):
+    session = settings.Session()
+
+    from airflow import models
```
Comment (feng-tao) [inline airflow/utils/db.py:291]: is there a reason we put `airflow import models` as local import not global import?

Comment (andscoop) [reply airflow/utils/db.py:291]: @feng-tao It is to prevent a circular import

**Review by tedmiston — COMMENTED** [review]
LGTM overall.  I think this PR is a nice enhancement for connections analogous to how `load_examples` works.

I have 2 comments for @andscoop and any other reviewers:

- Can you update the comment in default_airflow.cfg to explain why someone would want this setting on or off? (like how `load_examples` does)
- Does anyone think there are some connections we should always create even if this setting is false (e.g., the default `airflow_db`)?

**Comment by msumit** [comment]
@andscoop can you please add a couple of specs for the change or have a strong reason for not adding.

**Comment by ashb** [comment]
Isn't this what the `airflow upgradedb` command already does?

**Comment by ashb** [comment]
Additionally why _just_ the connections? And do we need a separate setting for this? Could we not just reuse the existing `load_examples` config option?

**Comment by msumit** [comment]
@ashb IMO examples and connections are 2 different entities altogether, and we should keep them different.

**Comment by ashb** [comment]
Can you think of a case where you'd want to load the Example DAGs but not the example connections? Or the inverse, when you want the example connections but not the example DAGs? (I never want either outside of a demo env)

**Comment by andscoop** [comment]
@ashb As it becomes easier to tear down and create an Airflow instance with dockerized solutions, init'ing without the example connections is preferable.  Beyond a users first time using airflow they will likely want to init without the example connections. `upgrade db` is a nice alternative if you have a long lived airflow instance that you have already manually removed the example conns from. This is a solution to avoid having to manually remove example conns.

I don't have a strong opinion on whether or not this should be combined with `load_examples`. The simplicity of combining them is always a plus, but I would need to see how users leverage the configs before speaking further.

**Comment by ashb** [comment]
> create an Airflow instance with dockerized solutions, init'ing without the example connections is preferable

`airflow upgradedb` already does that. It works from scratch too.

**Comment by andscoop** [comment]
@ashb I made an assumption that the [logic happening after the example connection creation](https://github.com/apache/incubator-airflow/blob/master/airflow/utils/db.py#L289) was necessary for a database initialization. For the most part, this does not appear to be the case.

I think this brings up the larger question, should the project continue to support both `airflow initdb` and `airflow upgradedb` in their current forms? It appears that `upgradedb` does the work a user expects out of `initdb` and that the extra logic in `initdb` is not necessary to initialization the database. 

ie.

[Creation of example chart](https://github.com/apache/incubator-airflow/blob/master/airflow/utils/db.py#L310)

[Pre-loading of DAG table](https://github.com/apache/incubator-airflow/blob/master/airflow/utils/db.py#L303)


There are two other core pieces of logic inside of `initdb` that I am unsure of. Is the [loading of the known event types](https://github.com/apache/incubator-airflow/blob/master/airflow/utils/db.py#L289) on database initialization necessary? It does not appear to be used beyond a single call at app creation time. If it is necessary the logic should be combined with `upgrade db`

I am also unsure of the [RBAC piece](https://github.com/apache/incubator-airflow/blob/master/airflow/utils/db.py#L328) inside of `initdb`.  If this is necessary we should combine with `upgrade db`.


I am happy to pivot this ticket into consolidation of these two methods, but would like to hear thoughts from maintainers first.

**Comment by ashb** [comment]
Since I didn't say so I think the goal of this ticket is 💯.

Ah the RBAC code there is an interesting one. Apart from that, everything in `initdb` is optional - airflow will operate fine with only ever running `upgradedb`. We have been running in prod since 1.8.2 only having done `upgradedb` and nothing has broken or missing (other than no sample charts or connections)

I have no idea where KnowEvent(Type) is actually used - there are very few references to it in the rest of the code base that I could see other the creation there, and a view. Nothing ever seems to use it for anything else.... Maybe a candidate for dropping, though worth checking with @mistercrunch or someone else (still?) at AirBnB.

Pre-loading of the DAG table didn't used to be required as starting the scheduler would do this job too, but on quick testing I'm not sure that's the case anymore.

Hmm, the RBAC change should possibly be converted to be an actual Alembic migration, or at the very least done via upgradedb not just initdb!

I have somewhat hijacked this issue though. Oops.

Having written al that I'm not sure the way forward either.

RBAC and KET should probably be in upgradedb. It might be worth merging them in to a single command behind a flag `airflow initdb --with-examples` which is more obvious/delibarate than having to know that `upgradedb` exists when all the tutorials just talk about initdb. THoughts?

**Comment by feluelle** [comment]
My opinion on this is that I can not initialize my database by using `airflow upgradedb` just because of the name. It does not make sense to me. I mean how can I `upgrade` a database when I did not even initialize it?!

> RBAC and KET should probably be in upgradedb. It might be worth merging them in to a single command behind a flag airflow initdb --with-examples which is more obvious/delibarate than having to know that upgradedb exists when all the tutorials just talk about initdb. THoughts?

I also dont know any case where I want to have example dags and not the default connections or vice versa. But I probably would not remove the possiblity to do that. Would you? :/

So in general `airflow initdb --with-examples` looks good to me but if you wanna be able to set both options seperatly it does not work.. :/

I think we should get more opinions on this.

**Comment by feng-tao** [comment]
Current RBAC models file are not available in airflow code base. I wonder whether we could do the alembic mgiration without the mode file.

**Comment by ashb** [comment]
The classes are always available (even when the rbac is not enabled) so we should be able to create migrations for them somehow.

Edit: thinking about it a bit more I'm not sure it makes sense to create Alembic migrations for the Flask RBAC models: if we import the model classes in the migration then the migrations can change over time (we got stung with this already once on our _own_ migrations), or we have to duplicate the model classes in the migration file, which means we have to manually keep them in sync/notice changes.

**Comment by hixus** [comment]
Any progress on this. I would prefer airflow initdb --with-examples which will init connections and example dags. Slight problem is that many operators use default_connection_x which doesn't really say that this is just a example connection. Also cloud-composer seems to be confiqured with default connections.

Anyway default connections makes it hard to see what connections are actually in use so not initializing them when examples are disabled would be a good thing.

**Pull request closed without merging**